### PR TITLE
Fix shadow rays and structure

### PR DESCRIPTION
**Pull request by novalain**

### CODE REVIEW

File: include/intersection_point.h [hunk 1]
```diff
@@ -9,10 +9,12 @@ class IntersectionPoint {
   Vertex position_;
   Direction normal_;
   Material material_;
+  float z_;
```
Comment (Carlbaum) [inline include/intersection_point.h:12]: we should probably rename this variable at some point

File: include/intersection_point.h [hunk 2]
```diff
@@ -9,10 +9,12 @@ class IntersectionPoint {
   Vertex position_;
```
Comment (Carlbaum) [inline include/intersection_point.h:9]: Why is this still left? Shouldn't float z replace it?!

File: include/intersection_point.h [hunk 3]
```diff
@@ -9,10 +9,12 @@ class IntersectionPoint {
...
   Direction normal_;
   Material material_;
+  float z_;
 public:
   IntersectionPoint();
-  IntersectionPoint(Vertex position, Direction normal, Material material_);
+  IntersectionPoint(Vertex position, Direction normal, Material material_, float z);
```
Comment (Carlbaum) [inline include/intersection_point.h:15]: Again, Vertex position should be deleted

File: include/triangle.h [hunk 4]
```diff
@@ -4,8 +4,9 @@
 #include "commons.h"
 #include "ray.h"
 #include "material.h"
+#include "scene_object.h"
 
-class Triangle {
+class Triangle : public SceneObject {
```
Comment (Carlbaum) [inline include/triangle.h:9]: Object oriented :S

File: src/geometry/sphere.cc [hunk 5]
```diff
@@ -24,22 +24,23 @@ bool Sphere::RayIntersection(Ray& ray, float& z) {
...
-  Vertex intersection_point = ray.get_origin() + ray.get_direction() * z;
+  Vertex intersection_point = ray.get_origin() + ray.get_direction() * t0;
   // TODO: This way should be equivalent to above, but it's not
   // Vertex intersection_point = position_ + radius_ * unit_normal;
   Direction normal = intersection_point - position_;
-  ray.set_intersection_point(
-      new IntersectionPoint(intersection_point, normal, material_));
-  return true;
+  //ray.set_intersection_point(
+    //  new IntersectionPoint(intersection_point, normal, material_));
```
Comment (Carlbaum) [inline src/geometry/sphere.cc]: These two lines can be removed?

File: src/geometry/sphere.cc [hunk 6]
```diff
@@ -24,22 +24,23 @@ bool Sphere::RayIntersection(Ray& ray, float& z) {
...
+
+  return new IntersectionPoint(intersection_point, normal, material_, t0);
+  //return true;
```
Comment (Carlbaum) [inline src/geometry/sphere.cc]: remove commented return

File: src/intersection_point.cc [hunk 7]
```diff
@@ -2,5 +2,6 @@
 
 IntersectionPoint::IntersectionPoint(Vertex position,
                                      Direction normal,
-                                     Material material)
-    : position_(position), normal_(normal), material_(material) {}
+                                     Material material,
+                                     float z)
+    : position_(position), normal_(normal), material_(material), z_(z) {}
```
Comment (Carlbaum) [inline src/intersection_point.cc:7]: Remove position. float z should replace it?

File: src/geometry/triangle.cc [hunk 8]
```diff
@@ -44,15 +44,17 @@ bool Triangle::RayIntersection(Ray& ray, float& z) {
...
-  if(u >= 0 && v >= 0 && u+v <= 1 && t > 0.f && t < z) { //if collision with a triangle closer to cam than before
+  if(u >= 0 && v >= 0 && u+v <= 1 && t > 0.f) { //if collision with a triangle closer to cam than before
     Vertex intersection_vertex = (1-u-v)*v0_ + u*v1_ + v*v2_;
-    ray.set_intersection_point(new IntersectionPoint(intersection_vertex, normal_, material_));
+    //ray.set_intersection_point(new IntersectionPoint(intersection_vertex, normal_, material_));
     // std::cout << "Successful update of pixel!" << std::endl; //TODO: remove when EVERYTHING is implemented
-    z = t;
-    return true;
+    //z = t;
+    //return true;
```
Comment (Carlbaum) [inline src/geometry/triangle.cc]: remove comments?

Comment (novalain) [reply src/geometry/triangle.cc]: whopps

File: include/intersection_point.h (hunk 1 above)
Comment (novalain) [reply include/intersection_point.h:12]: `z_distance_;`,  `t_near_;` ?

File: include/triangle.h (hunk 4 above)
Comment (novalain) [reply include/triangle.h:9]: Needed for `RayIntersection`

File: include/intersection_point.h (hunk 2 above)
Comment (novalain) [reply include/intersection_point.h:9]: No??? We still need intersection point position when calculating direction to light source? 
Or should we do ray.position + ray.direction*z every time :O ?

File: include/intersection_point.h (hunk 1 above)
Comment (Carlbaum) [reply include/intersection_point.h:12]: I think it's wrong to use z. And probably wrong to use t too.
Think just a plain "distance_" is enough.. or "distance_to_origin_" if you wanna be more accurate

File: include/triangle.h (hunk 4 above)
Comment (Carlbaum) [reply include/triangle.h:9]: yup, I know.. its just :S

**Comment by novalain** [comment]
Removed comments, can rebase now

**Review by Carlbaum — APPROVED** [review]

**Pull request merged**